### DISCUSSION TITLE
`try switch` statements

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#6759] [Language Ideas] [answered]
I think that a beneficial feature for C# would be to be able to mix exception handling and pattern matching in a single branching flow. This would take the form of a `try switch` statement, which would allow both `case` and `catch` arms. Here's an example:

```cs
    public void M() 
    {
        try switch (SomeFunction()) 
        {
            case null:
                Console.WriteLine("We have null"); 
                break;
            case {}:
                Console.WriteLine("We have an object"); 
                break;     
            catch MyException or MyOtherException:
                Console.WriteLine("We have a custom exception"); 
                break;
            catch Exception e:
                Console.WriteLine("We have some other exception");
                break;
        }
    }
```

which would desugar to a try/catch and two separate switch statements: one on the returned value of the expression and one on the caught exception:

```cs
    public void M() 
    {
        try 
        {
            switch (SomeFunction()) 
            {
                case null: 
                    Console.WriteLine("We have null"); 
                    break;
                case {}: 
                    Console.WriteLine("We have an object"); 
                    break;
            }
        }
        catch (Exception $ex) 
        {
            switch ($ex)
            {
                case MyException or MyOtherException:
                    Console.WriteLine("We have a custom Exception");
                    break;
                case Exception e
                    Console.WriteLine("We have some other exception");
                    break;
                default:
                    throw; // The catch-all behaviour (a ha) of exception handling is to rethrow, rather than getting a pattern match failure
            }
        }
    }
```

This has the following advantages:
1. reducing nesting (minor)
2. bringing the exception handling blocks closer to the code that threw the exception, aiding understanding (also see point 4)
3. being able to use proper pattern matching on the exceptions as well as on the returned value, such as: `catch CustomException { Reason: Reasons.Foo }: ...` (currently one must either put another switch in the catch block or try to use the limited features of exception catch guards)
4. it encourages wrapping the smallest unit of code in the `try`, which is normally good coding practice
5. lastly, this new form of control flow could also make its way to the expression form of switch, allowing the following code:

```cs
    public void M(string filename)
    {
        var fileContents =  
            OpenFile(filename) try switch
            {
                File file => file.Contents,
                catch FileNotFoundException => throw new ArgumentException(filename, "The requested file does not exist")
            };

        ... do something with fileContents ...
    }
```

whereas the current way you must write this has always left me somewhat unsatisfied, since you *must* split the declaration and initialisation of the `fileContents` variable:

```cs

    public void M(string filename)
    {
        String fileContents;

        try 
        {
            fileContents = OpenFile(filename);
        }
        catch (FileNotFoundException)
        {
            throw new ArgumentException(filename, "The requested file does not exist");
        }

        ... do something with fileContents ...
    }

```

The way you must write this method as above I believe encourages poorer coding styles, where the user feels encouraged to wrap the entire function in the try in order to write the code the way they would like.

As an added benefit, this would also give the language the oft-requested feature of "`try` as an expression", by writing an expression in the `catch` arms of the switch expression.

I'm fully open to arguments that this would lead to poor coding styles, so feel free to criticise; I did, however, think it was worthy of discussion.

**Comment by HaloFour** [comment, marked as answer]
Also see: #989 

If the C# language adds DUs and there is a BCL-"blessed" `Result<T, TError>` DU (or similar) with additional language support I could see getting something like this for free:

```cs
Result<string, Exception> result = try SomeExpression();

// lowered to something like
Result<string, Exception> result;
try {
    string $temp = SomeExpression();
    result = Result.Success<string, Exception>($temp);
}
catch (Exception exception) {
    result = Result.Failure<string, Exception>(exception);
}

// can be used like
switch (try SomeExpression()) {
    case Success(var value):
        // handle success
        break;
    case Failure(var reason):
        // handle failure
        break;
}
```

**Reply by Richiban** [reply]
Thanks @HaloFour, it makes sense something like this would have been suggested before! I like the interaction with the Result type; LanguageExt has something similar with the `Try(...)` wrapper type (although that obviously involves a lambda).
